docs/DataProviders.md

```diff
@@ -650,4 +586,268 @@ const UserProfile = ({ userId }) => {
+const convertFileToBase64 = file =>
+    new Promise((resolve, reject) => {
+        const reader = new FileReader();
+        reader.onload = () => resolve(reader.result);
+        reader.onerror = reject;
+        reader.readAsDataURL(file.rawFile);
+    });
+
+export default myDataProvider;
+```
+
+**Tip**: Use `beforeSave` instead of `beforeUpdate` to do the same for both create and update calls.
+
+### Sending Files In `multipart/form-data`
+
+In case you need to upload files, as you would with an HTML form but to your API, you can use the [FormData](https://developer.mozilla.org/en-US/docs/Web/API/FormData) API. It uses the same format as a form would use if the encoding type were set to `multipart/form-data`.
+
+The following `dataProvider` extends an existing one and tweaks its `create` and `update` methods for the `posts` resource only. It is the role of your API to negotiate the request and process the image. 
+
+The data provider sends the `post` data with the attached file in one query and works this way:
+- it checks if the resource is `posts`;
+- if so, it creates a new `FormData` object with the `post` data and the file received from the form;
+- it sends this `FormData` to the API with the react-admin [`fetchUtils.fetchJson()`](./fetchJson.md) function; 
+- if the resource is other than `posts`, it simply uses the `baseDataProvider` base methods.
+
+Let's have a look:
+
+```ts
+import simpleRestDataProvider from "ra-data-simple-rest";
+import {
+  CreateParams,
+  UpdateParams,
+  DataProvider,
+  fetchUtils,
+} from "react-admin";
+
+const endpoint = "http://path.to.my.api";
+const baseDataProvider = simpleRestDataProvider(endpoint);
+
+type PostParams = {
+  id: string;
+  title: string;
+  content: string;
+  picture: {
+    rawFile: File;
+    src?: string;
+    title?: string;
+  };
+};
+
+const createPostFormData = (
+  params: CreateParams<PostParams> | UpdateParams<PostParams>
+) => {
+  const formData = new FormData();
+  params.data.picture?.rawFile && formData.append("file", params.data.picture.rawFile);
+  params.data.title && formData.append("title", params.data.title);
+  params.data.content && formData.append("content", params.data.content);
+
+  return formData;
+};
+
+export const dataProvider: DataProvider = {
+  ...baseDataProvider,
+  create: (resource, params) => {
+    if (resource === "posts") {
+      const formData = createPostFormData(params);
+      return fetchUtils
+        .fetchJson(`${endpoint}/${resource}`, {
+          method: "POST",
+          body: formData,
+        })
+        .then(({ json }) => ({ data: json }));
+    }
+
+    return baseDataProvider.create(resource, params);
+  },
+  update: (resource, params) => {
+    if (resource === "posts") {
+      const formData = createPostFormData(params);
+      return fetchUtils
+        .fetchJson(`${endpoint}/${resource}`, {
+          method: "PUT",
+          body: formData,
+        })
+        .then(({ json }) => ({ data: json }));
+    }
```

Comment:
This example seems somewhat incomplete, as you never actually send the record id for instance. Don't know if that's a huge deal though, maybe we can still keep it that way since the 'essence' of the example (the `FormData`) is there.